elitzur-core/src/test/scala/com/spotify/elitzur/validators/DynamicRecordValidatorTest.scala

```diff
@@ -0,0 +1,140 @@
+ * under the License.
+ */
+
+package com.spotify.elitzur.validators
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import java.util.Locale
+
+import com.spotify.elitzur.MetricsReporter
+import com.spotify.elitzur.types.Owner
+import com.spotify.elitzur.validators.DynamicRecordValidatorTest.TestMetricsReporter
+
+case object Blizzard extends Owner {
+  override def name: String = "Blizzard"
+}
+
+object Companions {
+  implicit val nnlC: SimpleCompanionImplicit[Long, NonNegativeLong] =
+    SimpleCompanionImplicit(NonNegativeLongCompanion)
+}
+
+case class NonNegativeLong(data: Long) extends BaseValidationType[Long] {
+  override def checkValid: Boolean = data >= 0L
+}
+
+object NonNegativeLongCompanion extends BaseCompanion[Long, NonNegativeLong] {
+  def validationType: String = "NonNegativeLong"
+
+  def bigQueryType: String = "INTEGER"
+
+  def apply(data: Long): NonNegativeLong = NonNegativeLong(data)
+
+  def parse(data: Long): NonNegativeLong = NonNegativeLong(data)
+
+  override def owner: Owner = Blizzard
+
+  override def description: String = "Non negative long"
+}
+
+object DynamicRecordValidatorTest {
+  class TestMetricsReporter extends MetricsReporter {
+    val map : scala.collection.mutable.Map[String, Int] =
+      scala.collection.mutable.Map[String, Int]().withDefaultValue(0)
+    override def reportValid(className: String, fieldName: String, validationType: String): Unit =
+      map(s"$className.$fieldName.$validationType.valid") += 1
+    override def reportInvalid(className: String, fieldName: String, validationType: String): Unit =
+      map(s"$className.$fieldName.$validationType.invalid") += 1
+    override def toString: String = map.toString()
+    def getValid(className: String, fieldName: String, validationType: String): Int =
+      map(s"$className.$fieldName.$validationType.valid")
+    def getInvalid(className: String, fieldName: String, validationType: String): Int =
+      map(s"$className.$fieldName.$validationType.invalid")
+  }
+  def metricsReporter(): MetricsReporter = new TestMetricsReporter
+}
+
+
+class DynamicRecordValidatorTest extends AnyFlatSpec with Matchers {
+
+  it should "validate a simple case" in {
+    val label = "label"
+    val className = "com.spotify.DynamicClass"
+    implicit val metricsReporter: MetricsReporter = DynamicRecordValidatorTest.metricsReporter()
+    val nnlValidator = implicitly[Validator[NonNegativeLong]]
+    val recordValidator = DynamicRecordValidator(
+      Array(nnlValidator).asInstanceOf[Array[Validator[Any]]], Array(label))
+    recordValidator.validateRecord(
+      Unvalidated(Seq(NonNegativeLong(1L)).asInstanceOf[Seq[Any]]),
+      outermostClassName = Some(className)
+    )
+    metricsReporter.asInstanceOf[TestMetricsReporter].getValid(
+      className,
+      label,
+      NonNegativeLongCompanion.validationType
+    ) shouldEqual 1
+  }
+
+//  it should "validate a repeated value" in {
```

Comment:
I think this fails because of an unrelated bug with reporting metrics when validating repeated fields but need to verify